rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp

```diff
@@ -0,0 +1,220 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__LIFECYCLE_NODE_HPP_
+#define RCLCPP__LIFECYCLE_NODE_HPP_
+
+#include "rclcpp/node.hpp"
+#include "rclcpp_lifecycle/lifecycle_publisher.hpp"
+
+#include "rcl_lifecycle/lifecycle_state.h"
+#include "rcl_lifecycle/default_state_machine.h"
+#include "rcl_lifecycle/transition_map.h"
+
+namespace rclcpp
+{
+namespace node
+{
+
+namespace lifecycle_interface
+{
+
+/**
+ * @brief Interface class for a managed node.
+ * Pure virtual functions as defined in
+ * http://design.ros2.org/articles/node_lifecycle.html
+ */
+class NodeInterface
+{
+  virtual bool on_configure()   = 0;
+  virtual bool on_clean_up()    = 0;
+  virtual bool on_shutdown()    = 0;
+  virtual bool on_activate()    = 0;
+  virtual bool on_deactivate()  = 0;
+  virtual bool on_error()       = 0;
+};
+}  // namespace lifecycle_interface
+
+#include <rclcpp_lifecycle/visibility_control.h>
+
+/**
+ * @brief LifecycleNode as child class of rclcpp Node
+ * has lifecycle nodeinterface for configuring this node.
+ */
+class LifecycleNode : public rclcpp::node::Node, public lifecycle_interface::NodeInterface
+{
+public:
+
+  LIFECYCLE_EXPORT
+  explicit LifecycleNode(const std::string & node_name, bool use_intra_process_comms = false) :
+    Node(node_name, use_intra_process_comms)
+  {
+    setup_state_machine();
+  }
+
+  LIFECYCLE_EXPORT
+  ~LifecycleNode(){}
+
+  /**
+   * @brief get the default state machine
+   * as defined on design.ros.org
+   */
+  LIFECYCLE_EXPORT
+  virtual void
+  setup_state_machine()
+  {
+    state_machine_ = rcl_get_default_state_machine();
+  }
+
+  /**
+   * @brief same API for creating publisher as regular Node
+   */
+  template<typename MessageT, typename Alloc = std::allocator<void>>
+  std::shared_ptr<rclcpp::publisher::LifecyclePublisher<MessageT, Alloc>>
+  create_publisher(
+      const std::string & topic_name,
+      const rmw_qos_profile_t & qos_profile = rmw_qos_profile_default,
+      std::shared_ptr<Alloc> allocator = nullptr)
+  {
+    // create regular publisher in rclcpp::Node
+    auto pub =  rclcpp::node::Node::create_publisher<MessageT, Alloc,
+      rclcpp::publisher::LifecyclePublisher<MessageT, Alloc>>(
+        topic_name, qos_profile, allocator);
+
+    // keep weak handle for this publisher to enable/disable afterwards
+    weak_pub_ = pub;
+    return pub;
+  }
+
+  LIFECYCLE_EXPORT
+  void
+  print_state_machine()
+  {
+    printf("current state is %s\n", state_machine_.current_state.label);
+    rcl_print_transition_map(&state_machine_.transition_map);
+  }
+
+  LIFECYCLE_EXPORT
+  virtual bool
+  on_configure()
```

Comment:
Yes, the external API is what is missing. Your prototype's example gives the impression that you are treating the external and internal APIs the same.

I think that there should be a node interface that defines the external API. When the node is registered with a `LIfecycleManager` or something like that, then it would work with that API. A user could also manage their node's lifecycle directly from their main function, if they wish. Simultaneously, there needs to be a base class with virtual void methods (`on_activate` etc) for the implementer of a specific node to override to provide their own functionality.